### PR TITLE
interpret: use Either over Result when it is not representing an error condition

### PR DESCRIPTION
**Pull request by RalfJung**
r? @oli-obk

### CODE REVIEW

File: compiler/rustc_const_eval/src/interpret/operand.rs [hunk 1]
```diff
@@ -353,25 +355,25 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
                 if let Some(val) = self.read_immediate_from_mplace_raw(mplace)? {
-                    Ok(val)
+                    Left(val)
                 } else {
-                    Err(*mplace)
+                    Right(*mplace)
                 }
             }
-            Err(val) => Ok(val),
+            Right(val) => Left(val),
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/interpret/operand.rs:376]: May be useful to figure out a canonical left/right order for mplace vs valueish types

File: compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs [hunk 2]
```diff
@@ -147,7 +147,9 @@ impl IntRange {
                         // straight to the result, after doing a bit of checking. (We
                         // could remove this branch and just fall through, which
                         // is more general but much slower.)
-                        if let Ok(Ok(bits)) = scalar.to_bits_or_ptr_internal(target_size) {
+                        if let either::Left(bits) =
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs:150]: This whole if let could just be a `left` method call

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: compiler/rustc_const_eval/src/interpret/operand.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/operand.rs]: I swapped the order in what `read_immediate_raw` returns, the rest was already consistent.

**Review by oli-obk — APPROVED** [review]

**Pull request merged**